feat(STONEINTG-887): get base images from SBOM

MartinBasti opened the pull request:
Task `deprecated-image-check` will fetch data about base images also from SBOM.

This requires incompatible change, because users must update PLR definition with new parameters.

In future, konflux will stop provide BASE_IMAGES_DIGESTS via parameter, thus this param is optional now (we cannot remove it yet)

# Before you complete this pull request ...

Look for any open pull requests in the repository with the title "e2e-tests update" and 
see if there are recent e2e-tests updates that will be applicable to your change.

MartinBasti commented:
Example
```
STEP-CHECK-IMAGES

Fething SBOM for image quay.io/mbasti-test-org/single-nodejs-app:d1e1c90@sha256:1492539a932605755ed1bec583d5db4896828c37fdd0c61ca096485de8d9e27c
WARNING: Downloading SBOMs this way does not ensure its authenticity. If you want to ensure a tamper-proof SBOM, download it using 'cosign download attestation <image uri>' or verify its signature using 'cosign verify --key <key path> --attachment sbom <image uri>'.
Found SBOM of media type: application/vnd.cyclonedx+json
Detected base images from SBOM:

Base images pased by param BASE_IMAGES_DIGESTS: registry.access.redhat.com/ubi9/nodejs-16:<none>@sha256:c17111ec54c7f57f22d03f2abba206b0bdc54dcdfb02d6a8278ce088231eced1

Images to be checked:
registry.access.redhat.com/ubi9/nodejs-16

Querying Red Hat Catalog for registry.access.redhat.com/ubi9/nodejs-16.
Running conftest using /project/repository/ policy, required_checks namespace.
[
	{
		"filename": "/tmp/ubi9/nodejs-16/repository_data.json",
		"namespace": "required_checks",
		"successes": 1
	}
]
[SUCCESS] Image registry.access.redhat.com/ubi9/nodejs-16 is valid
{"result":"SUCCESS","timestamp":"1713542822","note":"Task deprecated-image-check completed: Check result for task result.","namespace":"required_checks","successes":1,"failures":0,"warnings":0}
```

MartinBasti commented:
Example when base images are available in SBOM:
```
STEP-CHECK-IMAGES

Fething SBOM for image quay.io/mbasti-test-org/single-container-app:f2566ab@sha256:e3605c5646ae1f6524b750b12a6c29fab309d9862491b1cee51f3a28202a5b63
WARNING: Downloading SBOMs this way does not ensure its authenticity. If you want to ensure a tamper-proof SBOM, download it using 'cosign download attestation <image uri>' or verify its signature using 'cosign verify --key <key path> --attachment sbom <image uri>'.
Found SBOM of media type: application/vnd.cyclonedx+json
Detected base images from SBOM:
docker.io/library/node

Base images pased by param BASE_IMAGES_DIGESTS: docker.io/library/node:latest@sha256:162d92c5f1467ad877bf6d8a098d9b04d7303879017a2f3644bfb1de1fc88ff0

Images to be checked:
docker.io/library/node

Querying Red Hat Catalog for docker.io/library/node.
[WARNING] Registry/image docker.io/library/node not found in Red Hat Catalog. Task cannot provide results if image is deprecated.
{"result":"WARNING","timestamp":"1713542891","note":"Task deprecated-image-check completed: Check result for task result.","namespace":"required_checks","successes":0,"failures":0,"warnings":1}
```

MartinBasti commented:
/retest

chmeliik commented on task/deprecated-image-check/0.4/deprecated-image-check.yaml:
```diff
@@ -0,0 +1,154 @@
+        # Get base images from SBOM
+        # images are without tag
+        echo "Fething SBOM for image ${IMAGE_URL}@${IMAGE_DIGEST}"
+        cosign download sbom "${IMAGE_URL}@${IMAGE_DIGEST}" > ${SBOM_FILE_PATH}
```
We were thinking other tasks could just read the SBOM from `<shared workspace root>/sbom-cyclonedx.json`

Although looking at this task, that would require a migration anyway :disappointed: 

In the future when tasks start using Trusted Artifacts, this task should switch from using an untrusted `cosign download` reference to using a trusted artifact reference. But that will be yet another migration and it probably doesn't matter much what you choose to do in the meantime

MartinBasti replied:
FBC workflow doesn't generate this file, but it had SBOM uploaded, it was working with previous version.
```
STEP-CHECK-IMAGES

[ERROR] SBOM not found at /workspace/workspace/sbom-cyclonedx.json.
{"result":"ERROR","timestamp":"1713803165","note":"Task deprecated-image-check failed: SBOM not found in the workspace.","namespace":"required_checks","successes":0,"failures":0,"warnings":0}
```

Other workflows are working, dunno why FBC is not working, `merge-syft-sboms` step in container-build taht generates this sbom should run for every workflow.


For example dockerfile workflow works fine:
```
STEP-CHECK-IMAGES

Detected base images from SBOM:
docker.io/library/node
```

MartinBasti replied:
FBC pipeline is special, `ls: cannot open directory '/workspace/workspace': Permission denied` no idea why :D 
Other pipelines have access

MartinBasti replied:
```
sh-5.1# ls -l
total 4
drwxr-xr-x. 5 root root 4096 Apr 22 17:11 workspace
sh-5.1# cd workspace/
sh: cd: workspace/: Permission denied
sh-5.1# id
uid=0(root) gid=0(root) groups=0(root)
```

the same permissions as the docker workflow

MartinBasti replied:
TLDR: issue was parallel access to workspace using EBS storage (it's not supported across nodes)

chmeliik commented on task/deprecated-image-check/0.4/deprecated-image-check.yaml:
```diff
@@ -0,0 +1,154 @@
+        # images are without tag
+        echo "Fething SBOM for image ${IMAGE_URL}@${IMAGE_DIGEST}"
+        cosign download sbom "${IMAGE_URL}@${IMAGE_DIGEST}" > ${SBOM_FILE_PATH}
+        cat "${SBOM_FILE_PATH}" | jq -r '.formulation? // empty | .[] | .components? // empty | .[] | select(.type=="container") | .name' > ${IMAGES_TO_BE_PROCESSED_PATH}
```
If you want to select exactly the base images that Konflux would report in BASE_IMAGES_DIGESTS, you can add

```jq
select(
  any((.properties // empty)[];
  .name | test("^konflux:container:is_(base|builder)_image"))
)
```

At the moment it only makes a hypothetical difference: what do you want deprecated-image-check to do if the `.formulation` part of the SBOM contains `container` type components reported by something other than Konflux?

MartinBasti replied:
updtated, thanks for the hint

MartinBasti commented:
/retest

jsztuka reviewed (approved):
lgtm

dirgim commented on task/deprecated-image-check/0.4/deprecated-image-check.yaml:
```diff
@@ -0,0 +1,155 @@
+      default: ""
+
+  results:
+    - name: PYXIS_HTTP_CODE
```
Do we still want to remove this result?

MartinBasti replied:
yes, but this code is not working due the permissions error, so I haven't removed it yet

MartinBasti replied:
that's because e2e tests are not testing it

MartinBasti commented:
Reverted to original code using cosign, we couldn't access workspace in parallel due to EBS storage.

Also removed unused result, to reuse version bump

dirgim reviewed (approved):
LGTM

hongweiliu17 commented on task/deprecated-image-check/0.4/MIGRATION.md:
```diff
@@ -0,0 +1,51 @@
+
+### New parameters
+
+Update files in Pull-Request created by RHTAP bot:
```
`Konflux bot`?

MartinBasti replied:
fixed

hongweiliu17 commented on task/deprecated-image-check/0.4/deprecated-image-check.yaml:
```diff
@@ -0,0 +1,150 @@
+
+        if [ -n "${BASE_IMAGES_DIGESTS}" ];
+        then
+          echo "Base images pased by param BASE_IMAGES_DIGESTS: $BASE_IMAGES_DIGESTS"
```
```suggestion
          echo "Base images passed by param BASE_IMAGES_DIGESTS: $BASE_IMAGES_DIGESTS"
```

MartinBasti replied:
fixed

hongweiliu17 reviewed (approved):
Two minor comment. Others look good to me.
